### PR TITLE
[feature](planner) Support  MetaScan for global dict

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: #9242

## Problem Summary:

This PR is a part of code of supports for global dict.

To support global dict, we need to have someway to let BE gathering and merge all the local dict to a global one, so when everytime the DictManager initial builds or updates a column dict,it will send a SQL like this: 

```sql
SELECT dict(d_date) FROM date [META];
```
As above sample shows, we added a new hint `META` and a new special aggregate function named `dict`.

In analyze phase, we will check if there is a hint named `META`, if there is, this stmt will be marked as a meta scan query.  For such stmt, we will generate MetaScanNode in Planner instead of OlapScanNode, and the column referenced by SlotRef in the function parameters will be added to MetaScanNode's `slotIdToDictId` field with Map type which will be used by BE to generate global dict, the value represents the dict_id which is always -1 for now. it wil be used for the incremental update of column dict in BE in the future. 

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    private final OlapTable olapTable;
+
+    public MetaScanNode(PlanNodeId id,
+                        TupleDescriptor tupleDesc) {
+        super(id, tupleDesc, NODE_NAME);
+        this.olapTable = (OlapTable) tupleDesc.getTable();
+    }
+
+    @Override
+    public List<TScanRangeLocations> getScanRangeLocations(long maxScanRangeLength) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java:76]: add some comment to explain why are parameters not used

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                scanRangeLocations.setScanRange(scanRange);
+
+                resultList.add(scanRangeLocations);
+            }
+        }
+        return resultList;
+    }
+
+    public void pullDictSlots(AggregateInfo aggInfo) {
+        Preconditions.checkState(aggInfo.getGroupingExprs().isEmpty());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java:140]: do we verify it and throw exception before this check？

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        List<FunctionCallExpr> funcExprList = aggInfo.getAggregateExprs();
+        for (FunctionCallExpr funcExpr : funcExprList) {
+            FunctionParams funcParams = funcExpr.getFnParams();
+            for (Expr expr : funcParams.exprs()) {
+                if (expr instanceof SlotRef && expr.isAnalyzed()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java:145]: when thie function called, expr should be analyzed, so we need to check `expr.isAnalyzed` ?

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java [hunk 4]
```diff
@@ -1159,6 +1159,10 @@ private PlanNode createAggregationPlan(SelectStmt selectStmt, Analyzer analyzer,
                     aggInfo.getSecondPhaseDistinctAggInfo());
             newRoot.init(analyzer);
             Preconditions.checkState(newRoot.hasValidStats());
+        } else if (aggInfo.isMetaScan()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java:1162]: we need to throw exception here

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java [hunk 5]
```diff
@@ -1683,6 +1687,11 @@ private PlanNode createScanNode(Analyzer analyzer, TableRef tblRef, SelectStmt s
 
         switch (tblRef.getTable().getType()) {
             case OLAP:
+                if (tblRef.isMetaScan()) {
+                    MetaScanNode metaScanNode = new MetaScanNode(ctx_.getNextNodeId(), tblRef.getDesc());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java:1691]: if we only could do meta scan on olap table, should we named this node to `OlapMeataScanNode` ?

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java [hunk 6]
```diff
@@ -1159,6 +1159,10 @@ private PlanNode createAggregationPlan(SelectStmt selectStmt, Analyzer analyzer,
                     aggInfo.getSecondPhaseDistinctAggInfo());
             newRoot.init(analyzer);
             Preconditions.checkState(newRoot.hasValidStats());
+        } else if (aggInfo.isMetaScan()) {
+            Preconditions.checkState(root instanceof MetaScanNode);
+            MetaScanNode metaScanNode = (MetaScanNode) root;
+            metaScanNode.pullDictSlots(aggInfo);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java:1165]: we need to check agg function name and then do `pullDictSlots`

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 7]
```diff
@@ -1310,6 +1310,8 @@ public void addBuiltinBothScalaAndVectorized(Function fn) {
 
     public static final String COUNT = "count";
     public static final String WINDOW_FUNNEL = "window_funnel";
+
+    public static final String DICT = "dict";
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 8]
```diff
@@ -341,6 +341,12 @@ public AggregateInfo getMergeAggInfo() {
 
     public boolean isMerge() { return aggPhase_.isMerge(); }
     public boolean isDistinctAgg() { return secondPhaseDistinctAggInfo_ != null; }
+
+    public boolean isMetaScan() {
+        return aggregateExprs_.size() > 0 &&
+            aggregateExprs_.get(0).getFn().getFunctionName().getFunction().equals(FunctionSet.DICT);
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java]: add check logic in the analyze phase of SelectStmt, if user do so, FE will throw a AnalysisException

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.planner;
+
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java:19]: done

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java (hunk 2 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java:141]: yes, we do!

File: fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java (hunk 3 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/MetaScanNode.java]: removed it

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java (hunk 6 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java:1165]: I think we not actually need check here currently,  have done this in the analyze phase

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java (hunk 5 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java]: done

File: fe/fe-core/src/test/java/org/apache/doris/analysis/CreateMaterializedViewStmtTest.java [hunk 10]
```diff
@@ -99,7 +99,7 @@ public void testCountDistinct(@Injectable SlotRef slotRef, @Injectable Arithmeti
         FunctionParams functionParams = new FunctionParams(true, fnChildren);
         FunctionCallExpr functionCallExpr = new FunctionCallExpr(FunctionSet.COUNT, functionParams);
         functionCallExpr.setFn(AggregateFunction.createBuiltin(FunctionSet.COUNT,
-                new ArrayList<>(), Type.BIGINT, Type.BIGINT, false, true, true));
+                new ArrayList<>(), Type.BIGINT, Type.BIGINT, false, true, false));
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/analysis/CreateMaterializedViewStmtTest.java:102]: why change that?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/SelectStmt.java [hunk 11]
```diff
@@ -1040,6 +1048,21 @@ private void analyzeAggregation(Analyzer analyzer) throws AnalysisException {
         aggExprs.clear();
         TreeNode.collect(substitutedAggs, Expr.isAggregatePredicate(), aggExprs);
 
+        if (metaQuery) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/SelectStmt.java:1051]: should we remove meta hint, and turn ScanNode to OlapMetaScanNode automaticaly when we found some function need to do meta query?

File: gensrc/thrift/PlanNodes.thrift [hunk 12]
```diff
@@ -52,6 +52,7 @@ enum TPlanNodeType {
   EXCEPT_NODE,
   ODBC_SCAN_NODE,
   TABLE_FUNCTION_NODE,
+  META_GATHER_NODE
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply gensrc/thrift/PlanNodes.thrift]: removed

File: fe/fe-core/src/test/java/org/apache/doris/analysis/CreateMaterializedViewStmtTest.java (hunk 10 above)
Comment (Kikyou1997) [reply fe/fe-core/src/test/java/org/apache/doris/analysis/CreateMaterializedViewStmtTest.java]: reverted

File: fe/fe-core/src/main/java/org/apache/doris/analysis/SelectStmt.java [hunk 13]
```diff
@@ -1040,6 +1048,21 @@ private void analyzeAggregation(Analyzer analyzer) throws AnalysisException {
...
+            long globalDictFuncCount = aggExprs.stream().filter(FunctionCallExpr::isGlobalDict).count();
+            if (globalDictFuncCount == 0) {
+                // simply ignore the hint
+                metaQuery = false;
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/SelectStmt.java:1055]: No.it wouldn't be created in that case

**Comment by Kikyou1997** [comment]
> please add unit test for it ❤️

Already committed in the newest patch

**Comment by morrySnow** [comment]
please fix conflicts ❤️

**Comment by Kikyou1997** [comment]
> please fix conflicts heart

done

**Comment by morrySnow** [comment]
fe-core compile failed in p0 regression, please fix it ~

**Comment by yiguolei** [comment]
duplicate with https://github.com/apache/incubator-doris/pull/9842

**Pull request closed without merging**